feat: 로그인, 로그아웃, 회원탈퇴 기능 구현.

ukjjang opened the pull request:
- 세션 기반 인증방식을 사용, 이후 Redis 세션 스토리지 적용 예정.
- LoginCheck AOP 적용.
- 컨트롤러에서 @CurrentUserId 파라미터로 받아올 수 있게 HandlerMethodArgumentResolver 설정 추가.
- jacoco, checkstyle gradle 파일 분리

f-lab-joseph commented on src/main/java/kr/bos/controller/UserController.java:
```diff
@@ -23,17 +27,54 @@
+    @ResponseStatus(HttpStatus.OK)
+    public void login(@Valid @RequestBody LoginInfoDto loginInfoDto) {
+        UserDto userDto = userService.selectUserByEmail(loginInfoDto.getEmail());
+        loginService.login(userDto, loginInfoDto.getPassword());
```
로그인 과정은 하나의 서비스 행위에 다 묶이는것이 좋지 않을까요 ? 로그인 서비스의 로직이 controller 구현에 의존적이게 된 것 같습니다만

ukjjang replied:
제가 작성한 login 과정은 크게 두 가지로 분리할 수 있다고 생각합니다.
1. 입력한 이메일로 UserDto를 받아오는 로직
2. 패스워드를 검증해서 session에 값을 추가하는 로직

사실 처음에는 로그인 loginService.login() 메소드 안에 두 가지 모두 포함시켰으나, 서비스의 로그인 메서드는 로그인 기능만 수행해야 한다고 생각했고,  1번 과정은 로그인이라는 기능이라기보다는 단순히 유저 정보를 가지고오는 기능이라는 생각을 들어 두 과정을 분리해야한다고 생각했습니다.

f-lab-joseph replied:
네, 말씀하신대로 로그인 과정을 분리할 수 있는데, 
서비스 영역은 앞단이 어떤 프로토콜이건 완결성있게 기능을 제공해줘야 하는데요. 
지금의 구조는 로그인 서비스가 로그인 기능을 완결성있게 제공해주지 않는 것 같아요 ㅎ

ukjjang replied:
그러네요.. userService.login()이 로그인 기능을 완결성있게 제공하고 있지 않네요.... 로그인 서비스에 로그인 과정이 모두 포함되는것이 맞는거 같습니다. 말씀대로 수정 작업하겠습니다!

f-lab-joseph commented on src/test/java/kr/bos/service/LoginServiceTest.java:
```diff
@@ -0,0 +1,73 @@
+    @DisplayName("로그인에 실패합니다. :잘못된 패스워드")
+    public void loginTestWhenFail() {
+        assertThrows(InvalidPasswordException.class,
+            () -> loginService.login(userDto, password + "@"));
```
빈값을 넣으면 어떻게 되나요 ?

f-lab-joseph replied:
임계값에 대한 테스트를 추가해주세요

ukjjang replied:
빈값, 해싱한 비밀번호를 추가해서 테스트 추가했습니다.

f-lab-joseph commented on src/main/java/kr/bos/service/LoginService.java:
```diff
@@ -0,0 +1,42 @@
+     */
+    public void login(UserDto userDto, String password) {
+        if (PasswordEncrypt.isMatch(password, userDto.getPassword())) {
+            session.setAttribute(USER_ID, userDto.getId());
```
서버가 이중화 됐을땐 어떻게할까요 ?

ukjjang replied:
이후 Redis 를 사용해 세션 클러스터링 처리할 예정입니다.

Commit:
feat: 로그인, 로그아웃, 회원탈퇴 기능 구현.
- 세션 기반 인증방식을 사용, 이후 Redis 세션 스토리지 적용 예정.
- LoginCheck AOP 적용.
- 컨트롤러에서 @CurrentUserId 파라미터로 받아올 수 있게 HandlerMethodArgumentResolver 설정 추가.
- jacoco, checkstyle gradle 파일 분리
